ikarus/utils/modalanalysis/modalanalysis.hh

```diff
@@ -0,0 +1,228 @@
+// SPDX-FileCopyrightText: 2021-2024 The Ikarus Developers mueller@ibb.uni-stuttgart.de
+// SPDX-License-Identifier: LGPL-3.0-or-later
+
+/**
+ * \file modalanalysis.hh
+ * \brief Implementation of modal analysis
+ */
+
+#pragma once
+#include <matplot/matplot.h>
+
+#include <dune/vtk/pvdwriter.hh>
+
+#include <Eigen/Core>
+#include <Eigen/SparseCore>
+
+#include <ikarus/assembler/assemblermanipulatorfuser.hh>
+#include <ikarus/assembler/simpleassemblers.hh>
+#include <ikarus/io/vtkwriter.hh>
+#include <ikarus/solver/eigenvaluesolver/generalizedeigensolver.hh>
+#include <ikarus/utils/concepts.hh>
+#include <ikarus/utils/modalanalysis/lumpingschemes.hh>
+#include <ikarus/utils/modalanalysis/modalanalysishelper.hh>
+#include <ikarus/utils/modalanalysis/tags.hh>
+
+namespace Ikarus::Dynamics {
+
+/**
+ * \brief Opinionated wrapper class for GeneralizedSymEigenSolver suited for modal analysis
+ *
+ * \tparam FEC the type of container for finite elements
+ * \tparam DV the type of the DirichletValues
+ */
+template <typename FEC, typename DV>
+struct ModalAnalysis
+{
+  using Assembler     = SparseFlatAssembler<FEC&, DV>;
+  using MatrixType    = typename Assembler::MatrixType;
+  using ScalarType    = typename MatrixType::Scalar;
+  using FERequirement = typename Assembler::FERequirement;
+  using FEContainer   = FEC;
+
+  using LumpedAssembler =
+      AssemblerManipulator<Assembler, Ikarus::Impl::AssemblerInterfaceHelper<ScalarAssembler, ScalarManipulator>,
+                           Ikarus::Impl::AssemblerInterfaceHelper<VectorAssembler, VectorManipulator>,
+                           Ikarus::Impl::AssemblerInterfaceHelper<MatrixAssembler, MatrixManipulator>>;
+  using Solver = GeneralizedSymEigenSolver<EigenValueSolverType::Spectra, MatrixType>;
+
+  /**
+   * \brief Construct a new Modal Analysis object
+   *
+   * \tparam FES deduced type of the container of the finite elements passed to the constructor
+   * \param fes the container of the finite elements
+   * \param dv the DirichletValues
+   */
+  template <typename FES>
+  ModalAnalysis(FES&& fes, const DV& dv)
+      : fes_(std::forward<FES>(fes)),
+        d_(dv.basis().size()),
+        dRef_(d_),
+        stiffAssembler_(std::make_shared<Assembler>(fes_, dv)),
+        massAssembler_(std::make_shared<Assembler>(fes_, dv)) {
+    if constexpr (std::remove_cvref_t<decltype(fes.front())>::Traits::useEigenRef)
+      req_.insertGlobalSolution(dRef_);
+    else
+      req_.insertGlobalSolution(d_);
+
+    stiffAssembler_->bind(req_, Ikarus::AffordanceCollections::elastoStatics, Ikarus::DBCOption::Reduced);
+    massAssembler_->bind(req_, Ikarus::AffordanceCollections::dynamics, Ikarus::DBCOption::Reduced);
+    lumpedMassAssembler_ = makeAssemblerManipulator(*massAssembler_);
+  }
+
+  /**
+   * \brief Binds a lumping scheme for the mass matrix. The Lumping scheme should have an operator () and modify the
```

Comment:
A concept can be added for the lumping scheme as well that it has to have an `operator()`?

Reply:
This is acually really tricky I think. 
You can check if it has a `operator()` but I think you have to specify the arguments, the first argument is the Assembler, which again is a template ...

Reply:
I did it with `std::is_invocable_v` maybe there is a better way